src/actions.rs

```diff
@@ -0,0 +1,886 @@
+        self.add_action(NullAction::Pause {
+            duration: duration.as_millis() as u64,
+        });
+    }
+
+    /// Add the specified action to this channel.
+    pub fn add_action(&mut self, action: NullAction) {
+        self.actions.push(action);
+    }
+}
+
+/// A channel containing `Key` actions.
+#[derive(Debug, Clone)]
+pub struct KeyActionChannel {
+    /// An identifier to distinguish this channel from others. Choose a meaningful string.
+    /// May be useful for debugging.
+    id: String,
+    /// The list of actions for this channel.
+    actions: Vec<KeyAction>,
+}
+
+impl KeyActionChannel {
+    /// Create a new Key channel.
+    pub fn new(id: &str) -> Self {
+        Self {
+            id: id.to_string(),
+            actions: Vec::new(),
+        }
+    }
+
+    /// Add a pause action to this channel.
+    pub fn add_pause(&mut self, duration: Duration) {
+        self.add_action(KeyAction::Pause {
+            duration: duration.as_millis() as u64,
+        });
+    }
+
+    /// Add the specified action to this channel.
+    pub fn add_action(&mut self, action: KeyAction) {
+        self.actions.push(action);
+    }
+}
+
+/// A channel containing `Key` actions.
+#[derive(Debug, Clone)]
+pub struct PointerActionChannel {
+    /// An identifier to distinguish this channel from others. Choose a meaningful string.
+    /// May be useful for debugging.
+    id: String,
+    /// The pointer type. Can be `Mouse`, `Pen` or `Touch`.
+    pointer_type: PointerActionType,
+    /// The list of actions for this channel.
+    actions: Vec<PointerAction>,
+}
+
+impl PointerActionChannel {
+    /// Create a new Pointer channel.
+    pub fn new(id: &str, pointer_type: PointerActionType) -> Self {
+        Self {
+            id: id.to_string(),
+            pointer_type,
+            actions: Vec::new(),
+        }
+    }
+
+    /// Add a pause action to this channel.
+    pub fn add_pause(&mut self, duration: Duration) {
+        self.add_action(PointerAction::Pause {
+            duration: duration.as_millis() as u64,
+        });
+    }
+
+    /// Add the specified action to this channel.
+    pub fn add_action(&mut self, action: PointerAction) {
+        self.actions.push(action);
+    }
+}
+
+/// An ActionChannel is a sequence of actions of a specific type.
+/// When combined with other channels, you can think of them like a grid, with actions on the
+/// horizontal axis and channels on the vertical axis. All of the actions in the first column
+/// will be executed simultaneously, then all of the second actions, and so on.
+/// The second column will not be executed until all actions in the first column have been
+/// completed, including any pauses.
+///
+/// Thus, the duration of each column will be equal to the longest duration of any individual
+/// action in that column.
+#[derive(Debug, Clone)]
+pub enum ActionChannel {
+    /// A channel containing null actions.
+    Null(NullActionChannel),
```

Comment:
Is the "Null" naming dictated by the spec? Feels like a very non-descript word for what these actions really represent.

Reply:
It's used in both the python selenium library as well as the `webdriver` types.
They also use `GeneralAction` for the pause action so that might be nicer?

Reply:
I think it'd be helpful to link to the relevant spec section for each input source type.